kolibri/plugins/facility/assets/src/routes.js

```diff
@@ -41,7 +43,14 @@ export default [
     path: '/:facility_id?/classes',
     component: ManageClassPage,
     handler: toRoute => {
-      showClassesPage(store, toRoute);
+      if (store.getters.userIsMultiFacilityAdmin && !toRoute.params.facility_id) {
```

Comment:
I wonder if we could clean this up a bit w/ a [Vue Router Guard](https://router.vuejs.org/guide/advanced/navigation-guards.html) like a

```javascript
router.beforeEach((to, from) => {
    if(store.getters... && !to.params.facility_id) { 
         return router.replace({...});
    } else {
        return true; // indicates that things should just move along
    }
})
```

Reply:
can you say a bit more about where you imagine this being placed and elaborate a bit? i totally agree with making it more DRY, but tried a lot of different things to get this working so would love to hear more detail about what you're thinking 🙂 

it's my understanding that the handlers in each of these routes are [effectively acting as a beforeEach](https://github.com/learningequality/kolibri/blob/develop/kolibri/core/assets/src/router.js#L50) for each route, and i wasn't successful when trying to place many variations of this logic in, for example, a more global `beforeEach` within `ready()` in facility's `app.js`. i have tried (and failed at haha) rendering this logic more extensible and/or general and also placing it in a lot of different areas of the plugin, so at this point i'm hoping for more specific guidance, as my ideas are all tapped out. 

i also thought through the possibility of adding something like a `showAllFacilitiesPage` handler which could be checked (and passed a variable or two) before then turning to the page-specific handler, but right now each of the routes with their own handlers are their own modules while `AllFacilitesPage` is just a stand-alone file, and there didn't seem like there was structure in place to support that change nor that it was worth it to refactor the file and split it apart.

Reply:
Not sure if [this](https://github.com/learningequality/kolibri/blob/3937057d00c37d1f1db6fae74461aa3686446879/kolibri/plugins/device/assets/src/app.js#L56) will help, but I recently implemented a nav guard for the pin setup feature I was working in. The only difference is that I was implementing this within the `beforeResolve`

Reply:
yes, this is actually precisely where i initially tried to implement this, @akolson & @nucleogenesis! but it ended up introducing even more "stuck navigation" errors like the one mentioned in the `TODO` in the description, #10649. i'll revisit it since i've changed a couple things since then, and ping you again once i have either updated it or have a better explanation for not using it than my current fuzzy memory 🙂 thanks for your feedback!!

Reply:
I don't think it's super important to clean this up so no need to go down this route, but I think that maybe a wildcard route could work like:

```
  {
    path: '*',
    beforeEnter: (toRoute, fromRoute, next) => {
      if (store.getters.userIsMultiFacilityAdmin && !toRoute.params.facility_id) {
        next({
          name: 'ALL_FACILITIES_PAGE',
          // not sure what the `subTopicName` param does though so this might not work because of that?
        });
      } else {
        next(); // Should move along to the `toRoute` I don't think you have to pass `toRoute` to `next()`
      }
    },
  },

```

In any case, just a thought and I think it's something that can be put off as a dev chore to look into it for a patch release or something.

Reply:
I think the main reason for not DRYing this up is that it's not a check that should happen before _every_ route - for example, we don't want this to happen before the `ALL_FACILITIES_PAGE` route itself.

One way it could be dried up slightly would be to create a function that does this check, and then invoke it in each handler function, similar to the unassigned content guard in Learn: https://github.com/learningequality/kolibri/blob/develop/kolibri/plugins/learn/assets/src/routes/index.js#L25

So, in the case that it catches, it returns `true`. E.g.:

```
function facilityParamRequiredGuard(toRoute, next, subtopicName) {
  if (store.getters.userIsMultiFacilityAdmin && !toRoute.params.facility_id) {
    next({
        name: 'ALL_FACILITIES_PAGE',
        params: { subtopicName },
      });
      return true;
  }
}
```

then the handler function in this case simplifies to:
```
handler: (toRoute, fromRoute, next) => {
  if (facilityParamRequiredGuard(toRoute, next, ManageClassPage.name)) {
    return;
  }
  showClassesPage(store, toRoute);
},
```

The other thing this does is use the [return early pattern](https://gomakethings.com/the-early-return-pattern-in-javascript/), which helps to reduce the nesting in functions.

Reply:
thanks @rtibbles! this is very close to what i was envisioning as behavior for a `showAllFacilitiesPage` handler that gets called in the needed routes before hitting the page-specific handler, but i wasn't really sure how it would fit into the current code structure, which this completely clarifies. i'll be trying it out now and should be updating soon!